File: src/MyServiceBus.RabbitMq/RabbitMqTransportFactory.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Threading.Tasks;
using MyServiceBus.Topology;
using RabbitMQ.Client;

namespace MyServiceBus;

public sealed class RabbitMqTransportFactory : ITransportFactory
{
    private readonly ConnectionProvider _connectionProvider;
    private readonly ConcurrentDictionary<string, ISendTransport> _sendTransports = new();
    private readonly ConcurrentDictionary<string, ISendTransport> _queueTransports = new();
    private readonly ushort _prefetchCount;
    private readonly Uri _baseAddress;

    public TransportCapabilityDescriptor Capabilities => TransportCapabilityDescriptors.RabbitMq;

    public RabbitMqTransportFactory(ConnectionProvider connectionProvider, IRabbitMqFactoryConfigurator configurator)
    {
        _connectionProvider = connectionProvider;
        _prefetchCount = configurator.PrefetchCount;
        _baseAddress = new UriBuilder("rabbitmq", configurator.ClientHost, configurator.ClientPort).Uri;
    }

    public Uri GetPublishAddress(string entityName) => new(_baseAddress, $"exchange/{entityName}");

    public Uri GetTemporaryEndpointAddress(string endpointName) =>
        new(_baseAddress, $"exchange/{endpointName}?durable=false&autodelete=true");

    public Uri GetErrorAddress(string endpointName) => GetPublishAddress(endpointName + "_error");

    public Uri GetFaultAddress(string endpointName) => GetPublishAddress(endpointName + "_fault");

    public async Task<ISendTransport> GetSendTransport(Uri address, CancellationToken cancellationToken = default)
    {
        string? exchange = null;
        string? queue = null;
        bool durable = true;
        bool autoDelete = false;

        if (address.Scheme.Equals("exchange", StringComparison.OrdinalIgnoreCase))
        {
            var spec = address.OriginalString.Substring("exchange:".Length);
            string? query = null;
            var idx = spec.IndexOf('?');
            if (idx >= 0)
            {
                query = spec[(idx + 1)..];
                spec = spec[..idx];
            }

            exchange = spec;
            ParseExchangeSettings(query, ref durable, ref autoDelete);
        }
        else if (address.Scheme.Equals("queue", StringComparison.OrdinalIgnoreCase))
        {
            var spec = address.OriginalString.Substring("queue:".Length);
            string? query = null;
            var idx = spec.IndexOf('?');
            if (idx >= 0)
            {
                query = spec[(idx + 1)..];
                spec = spec[..idx];
            }

            queue = spec;
            ParseExchangeSettings(query, ref durable, ref autoDelete);
        }
        else
        {
            var path = address.AbsolutePath.Trim('/');
            if (path.StartsWith("exchange/", StringComparison.OrdinalIgnoreCase))
            {
                exchange = path["exchange/".Length..];
            }
            else if (!string.IsNullOrEmpty(path))
            {
                exchange = path;
            }

            ParseExchangeSettings(address.Query.TrimStart('?'), ref durable, ref autoDelete);

            if (exchange == null && queue == null)
                exchange = "default";
        }

        if (queue != null)
        {
            if (!_queueTransports.TryGetValue(queue, out var queueTransport))
            {
                var connection = await _connectionProvider.GetOrCreateConnectionAsync(cancellationToken);
                var channel = await connection.CreateChannelAsync(cancellationToken: cancellationToken);

                if (!autoDelete)
                {
                    var errorExchange = queue + "_error";
                    var errorQueue = errorExchange;
                    var skippedExchange = queue + "_skipped";
                    var skippedQueue = skippedExchange;

                    await channel.ExchangeDeclareAsync(
                        exchange: errorExchange,
                        type: ExchangeType.Fanout,
                        durable: durable,
                        autoDelete: autoDelete,
                        cancellationToken: cancellationToken);

                    await channel.QueueDeclareAsync(
                        queue: errorQueue,
                        durable: durable,
                        exclusive: false,
                        autoDelete: autoDelete,
                        cancellationToken: cancellationToken);

                    await channel.QueueBindAsync(
                        queue: errorQueue,
                        exchange: errorExchange,
                        routingKey: string.Empty,
                        cancellationToken: cancellationToken);

                    await channel.ExchangeDeclareAsync(
                        exchange: skippedExchange,
                        type: ExchangeType.Fanout,
                        durable: durable,
                        autoDelete: autoDelete,
                        cancellationToken: cancellationToken);

                    await channel.QueueDeclareAsync(
                        queue: skippedQueue,
                        durable: durable,
                        exclusive: false,
                        autoDelete: autoDelete,
                        cancellationToken: cancellationToken);

                    await channel.QueueBindAsync(
                        queue: skippedQueue,
                        exchange: skippedExchange,
                        routingKey: string.Empty,
                        cancellationToken: cancellationToken);

                    var faultExchange = queue + "_fault";
                    var faultQueue = faultExchange;

                    await channel.ExchangeDeclareAsync(
                        exchange: faultExchange,
                        type: ExchangeType.Fanout,
                        durable: durable,
                        autoDelete: autoDelete,
                        cancellationToken: cancellationToken);

                    await channel.QueueDeclareAsync(
                        queue: faultQueue,
                        durable: durable,
                        exclusive: false,
                        autoDelete: autoDelete,
                        cancellationToken: cancellationToken);

                    await channel.QueueBindAsync(
                        queue: faultQueue,
                        exchange: faultExchange,
                        routingKey: string.Empty,
                        cancellationToken: cancellationToken);
                }

                await channel.QueueDeclareAsync(
                    queue: queue,
                    durable: durable,
                    exclusive: false,
                    autoDelete: autoDelete,
                    cancellationToken: cancellationToken);

                queueTransport = new RabbitMqQueueSendTransport(channel, queue);
                _queueTransports.TryAdd(queue, queueTransport);
            }

            return queueTransport;
        }

        exchange ??= "default";
        if (!_sendTransports.TryGetValue(exchange, out var sendTransport))
        {
            var connection = await _connectionProvider.GetOrCreateConnectionAsync(cancellationToken);
            var channel = await connection.CreateChannelAsync(cancellationToken: cancellationToken);
            await channel.ExchangeDeclareAsync(exchange, type: ExchangeType.Fanout, durable: durable, autoDelete: autoDelete, cancellationToken: cancellationToken);
            sendTransport = new RabbitMqSendTransport(channel, exchange);

            _sendTransports.TryAdd(exchange, sendTransport);
        }

        return sendTransport;
    }

    public async Task<IReceiveTransport> CreateReceiveTransport(
        ReceiveEndpointTopology topology,
        Func<ReceiveContext, Task> handler,
        Func<string?, bool>? isMessageTypeRegistered = null,
        CancellationToken cancellationToken = default)
    {
        var rabbitMqTopology = RabbitMqReceiveEndpointTopology.Project(topology);
        var connection = await _connectionProvider.GetOrCreateConnectionAsync(cancellationToken);
        var channel = await connection.CreateChannelAsync(cancellationToken: cancellationToken);

        var prefetch = rabbitMqTopology.PrefetchCount > 0 ? rabbitMqTopology.PrefetchCount : _prefetchCount;
        if (prefetch > 0)
            await channel.BasicQosAsync(0, prefetch, false, cancellationToken);

        await channel.ExchangeDeclareAsync(
            exchange: rabbitMqTopology.ExchangeName,
            type: rabbitMqTopology.ExchangeType,
            durable: rabbitMqTopology.Durable,
            autoDelete: rabbitMqTopology.AutoDelete,
            cancellationToken: cancellationToken
        );

        var hasErrorQueue = !rabbitMqTopology.AutoDelete;

        if (hasErrorQueue)
        {
            var errorExchange = rabbitMqTopology.QueueName + "_error";
            var errorQueue = errorExchange;
            var skippedExchange = rabbitMqTopology.QueueName + "_skipped";
            var skippedQueue = skippedExchange;

            await channel.ExchangeDeclareAsync(
                exchange: errorExchange,
                type: ExchangeType.Fanout,
                durable: rabbitMqTopology.Durable,
                autoDelete: rabbitMqTopology.AutoDelete,
                cancellationToken: cancellationToken
            );

            await channel.QueueDeclareAsync(
                queue: errorQueue,
                durable: rabbitMqTopology.Durable,
                exclusive: false,
                autoDelete: rabbitMqTopology.AutoDelete,
                cancellationToken: cancellationToken
            );

            await channel.QueueBindAsync(
                queue: errorQueue,
                exchange: errorExchange,
                routingKey: string.Empty,
                cancellationToken: cancellationToken
            );

            await channel.ExchangeDeclareAsync(
                exchange: skippedExchange,
                type: ExchangeType.Fanout,
                durable: rabbitMqTopology.Durable,
                autoDelete: rabbitMqTopology.AutoDelete,
                cancellationToken: cancellationToken
            );

            await channel.QueueDeclareAsync(
                queue: skippedQueue,
                durable: rabbitMqTopology.Durable,
                exclusive: false,
                autoDelete: rabbitMqTopology.AutoDelete,
                cancellationToken: cancellationToken
            );

            await channel.QueueBindAsync(
                queue: skippedQueue,
                exchange: skippedExchange,
                routingKey: string.Empty,
                cancellationToken: cancellationToken
            );

            var faultExchange = rabbitMqTopology.QueueName + "_fault";
            var faultQueue = faultExchange;

            await channel.ExchangeDeclareAsync(
                exchange: faultExchange,
                type: ExchangeType.Fanout,
                durable: rabbitMqTopology.Durable,
                autoDelete: rabbitMqTopology.AutoDelete,
                cancellationToken: cancellationToken
            );

            await channel.QueueDeclareAsync(
                queue: faultQueue,
                durable: rabbitMqTopology.Durable,
                exclusive: false,
                autoDelete: rabbitMqTopology.AutoDelete,
                cancellationToken: cancellationToken
            );

            await channel.QueueBindAsync(
                queue: faultQueue,
                exchange: faultExchange,
                routingKey: string.Empty,
                cancellationToken: cancellationToken
            );
        }

        await channel.QueueDeclareAsync(
            queue: rabbitMqTopology.QueueName,
            durable: rabbitMqTopology.Durable,
            exclusive: false,
            autoDelete: rabbitMqTopology.AutoDelete,
            arguments: rabbitMqTopology.QueueArguments,
            cancellationToken: cancellationToken
        );

        await channel.QueueBindAsync(
            queue: rabbitMqTopology.QueueName,
            exchange: rabbitMqTopology.ExchangeName,
            routingKey: rabbitMqTopology.RoutingKey,
            cancellationToken: cancellationToken
        );

        var errorAddress = hasErrorQueue ? GetErrorAddress(rabbitMqTopology.QueueName) : null;
        var faultAddress = hasErrorQueue ? GetFaultAddress(rabbitMqTopology.QueueName) : null;
        return new RabbitMqReceiveTransport(channel, rabbitMqTopology.QueueName, handler, errorAddress, faultAddress, isMessageTypeRegistered);
    }

    private static void ParseExchangeSettings(string? queryString, ref bool durable, ref bool autoDelete)
    {
        if (string.IsNullOrEmpty(queryString))
            return;

        var query = queryString.Split('&', StringSplitOptions.RemoveEmptyEntries);
        foreach (var part in query)
        {
            var kv = part.Split('=', 2, StringSplitOptions.RemoveEmptyEntries);
            if (kv.Length != 2)
                continue;

            var key = kv[0];
            var value = kv[1];

            if (key.Equals("durable", StringComparison.OrdinalIgnoreCase) && bool.TryParse(value, out var d))
                durable = d;
            else if (key.Equals("autodelete", StringComparison.OrdinalIgnoreCase) && bool.TryParse(value, out var ad))
                autoDelete = ad;
            else if (key.Equals("temporary", StringComparison.OrdinalIgnoreCase) && bool.TryParse(value, out var temporary) && temporary)
            {
                durable = false;
                autoDelete = true;
            }
        }
    }
}

File: src/Java/myservicebus-rabbitmq/src/main/java/com/myservicebus/rabbitmq/RabbitMqTransportFactory.java
package com.myservicebus.rabbitmq;

import java.net.URI;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

import com.myservicebus.ReceiveTransport;
import com.myservicebus.SendTransport;
import com.myservicebus.TransportFactory;
import com.myservicebus.TransportCapabilityDescriptor;
import com.myservicebus.TransportCapabilityDescriptors;
import com.myservicebus.TransportMessage;
import com.myservicebus.topology.MessageBinding;
import com.myservicebus.topology.ReceiveEndpointTransportTopology;
import com.rabbitmq.client.BuiltinExchangeType;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.myservicebus.logging.LoggerFactory;

public class RabbitMqTransportFactory implements TransportFactory {
    @Override
    public TransportCapabilityDescriptor getCapabilities() {
        return TransportCapabilityDescriptors.RABBITMQ;
    }

    private final ConnectionProvider connectionProvider;
    private final ConcurrentHashMap<String, RabbitMqSendTransport> exchangeTransports = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<String, RabbitMqSendTransport> queueTransports = new ConcurrentHashMap<>();
    private final int defaultPrefetchCount;
    private final LoggerFactory loggerFactory;

    public RabbitMqTransportFactory(ConnectionProvider connectionProvider, RabbitMqFactoryConfigurator configurator,
            LoggerFactory loggerFactory) {
        this.connectionProvider = connectionProvider;
        this.defaultPrefetchCount = configurator.getPrefetchCount();
        this.loggerFactory = loggerFactory;
    }

    public SendTransport getSendTransport(String exchange, boolean durable, boolean autoDelete) {
        String key = exchange + ":" + durable + ":" + autoDelete;
        return exchangeTransports.computeIfAbsent(key, ex -> {
            try {
                Connection connection = connectionProvider.getOrCreateConnection();
                Channel channel = connection.createChannel();
                channel.exchangeDeclare(exchange, "fanout", durable, autoDelete, null);
                return new RabbitMqSendTransport(channel, exchange, "");
            } catch (Exception e) {
                throw new RuntimeException("Failed to create send transport", e);
            }
        });
    }

    public SendTransport getQueueTransport(String queue) {
        return getQueueTransport(queue, true, false);
    }

    public SendTransport getQueueTransport(String queue, boolean durable, boolean autoDelete) {
        String key = queue + ":" + durable + ":" + autoDelete;
        return queueTransports.computeIfAbsent(key, q -> {
            try {
                Connection connection = connectionProvider.getOrCreateConnection();
                Channel channel = connection.createChannel();

                if (!autoDelete) {
                    String errorExchange = queue + "_error";
                    String errorQueue = queue + "_error";
                    String skippedExchange = queue + "_skipped";
                    String skippedQueue = queue + "_skipped";
                    String faultExchange = queue + "_fault";
                    String faultQueue = queue + "_fault";

                    channel.exchangeDeclare(errorExchange, "fanout", durable, autoDelete, null);
                    channel.queueDeclare(errorQueue, durable, false, autoDelete, null);
                    channel.queueBind(errorQueue, errorExchange, "");

                    channel.exchangeDeclare(skippedExchange, "fanout", durable, autoDelete, null);
                    channel.queueDeclare(skippedQueue, durable, false, autoDelete, null);
                    channel.queueBind(skippedQueue, skippedExchange, "");

                    channel.exchangeDeclare(faultExchange, "fanout", durable, autoDelete, null);
                    channel.queueDeclare(faultQueue, durable, false, autoDelete, null);
                    channel.queueBind(faultQueue, faultExchange, "");
                }

                channel.queueDeclare(queue, durable, false, autoDelete, null);

                return new RabbitMqSendTransport(channel, "", queue);
            } catch (Exception e) {
                throw new RuntimeException("Failed to create send transport", e);
            }
        });
    }

    @Override
    public SendTransport getSendTransport(URI address) {
        if ("exchange".equalsIgnoreCase(address.getScheme())) {
            return getSendTransport(address.getSchemeSpecificPart(), true, false);
        }
        if ("queue".equalsIgnoreCase(address.getScheme())) {
            return getQueueTransport(address.getSchemeSpecificPart(), true, false);
        }

        String path = address.getPath();
        boolean temporary = queryFlag(address, "temporary");
        boolean durable = !temporary && !queryFlag(address, "durable", false);
        boolean autoDelete = temporary || queryFlag(address, "autodelete");
        if (path.contains("/exchange/")) {
            String exchange = path.substring(path.lastIndexOf('/') + 1);
            return getSendTransport(exchange, durable, autoDelete);
        }
        String exchange = path.substring(1);
        return getSendTransport(exchange, durable, autoDelete);
    }

    private static boolean queryFlag(URI address, String name) {
        return queryFlag(address, name, true);
    }

    private static boolean queryFlag(URI address, String name, boolean expectedValue) {
        String query = address.getQuery();
        if (query == null || query.isBlank()) {
            return false;
        }

        for (String part : query.split("&")) {
            String[] pair = part.split("=", 2);
            if (pair.length == 2 && pair[0].equalsIgnoreCase(name)) {
                return Boolean.parseBoolean(pair[1]) == expectedValue;
            }
        }
        return false;
    }

    @Override
    public ReceiveTransport createReceiveTransport(String queueName, List<MessageBinding> bindings,
            Function<TransportMessage, CompletableFuture<Void>> handler,
            Function<String, Boolean> isMessageTypeRegistered, int prefetchCount) throws Exception {
        return createReceiveTransport(queueName, bindings, handler, isMessageTypeRegistered, prefetchCount, null);
    }

    @Override
    public ReceiveTransport createReceiveTransport(String queueName, List<MessageBinding> bindings,
            Function<TransportMessage, CompletableFuture<Void>> handler,
            Function<String, Boolean> isMessageTypeRegistered, int prefetchCount,
            Map<String, Object> queueArguments) throws Exception {
        RabbitMqReceiveEndpointTopology topology = RabbitMqReceiveEndpointTopology.project(
                queueName, bindings, prefetchCount, queueArguments);
        return createReceiveTransport(topology, handler, isMessageTypeRegistered);
    }

    @Override
    public ReceiveTransport createReceiveTransport(ReceiveEndpointTransportTopology topology,
            Function<TransportMessage, CompletableFuture<Void>> handler,
            Function<String, Boolean> isMessageTypeRegistered) throws Exception {
        return createReceiveTransport(
                RabbitMqReceiveEndpointTopology.project(topology), handler, isMessageTypeRegistered);
    }

    private ReceiveTransport createReceiveTransport(RabbitMqReceiveEndpointTopology topology,
            Function<TransportMessage, CompletableFuture<Void>> handler,
            Function<String, Boolean> isMessageTypeRegistered) throws Exception {
        Connection connection = connectionProvider.getOrCreateConnection();
        Channel channel = connection.createChannel();

        int count = topology.prefetchCount() > 0 ? topology.prefetchCount() : defaultPrefetchCount;
        if (count > 0) {
            channel.basicQos(count);
        }

        for (MessageBinding binding : topology.bindings()) {
            String exchangeName = binding.getEntityName();
            channel.exchangeDeclare(exchangeName, BuiltinExchangeType.FANOUT, true);
            channel.queueDeclare(topology.queueName(), true, false, false, topology.queueArguments());
            channel.queueBind(topology.queueName(), exchangeName, "");
        }

        String errorExchange = topology.queueName() + "_error";
        String errorQueue = errorExchange;
        channel.exchangeDeclare(errorExchange, BuiltinExchangeType.FANOUT, true);
        channel.queueDeclare(errorQueue, true, false, false, null);
        channel.queueBind(errorQueue, errorExchange, "");

        String skippedExchange = topology.queueName() + "_skipped";
        String skippedQueue = skippedExchange;
        channel.exchangeDeclare(skippedExchange, BuiltinExchangeType.FANOUT, true);
        channel.queueDeclare(skippedQueue, true, false, false, null);
        channel.queueBind(skippedQueue, skippedExchange, "");
        String faultExchange = topology.queueName() + "_fault";
        String faultQueue = faultExchange;
        channel.exchangeDeclare(faultExchange, BuiltinExchangeType.FANOUT, true);
        channel.queueDeclare(faultQueue, true, false, false, null);
        channel.queueBind(faultQueue, faultExchange, "");

        String faultAddress = getFaultAddress(topology.queueName());
        return new RabbitMqReceiveTransport(channel, topology.queueName(), handler, faultAddress, isMessageTypeRegistered,
                loggerFactory);
    }

    @Override
    public String getPublishAddress(String exchange) {
        return connectionProvider.getPublishAddress(exchange);
    }

    @Override
    public String getSendAddress(String queue) {
        return connectionProvider.getSendAddress(queue);
    }
}
